cbmc_viewer/aliast.py

```diff
@@ -0,0 +1,182 @@
+    })
+
+VALID_ALIAS_SUMMARY = voluptuous.schema_builder.Schema({
+    'summary': VALID_ALIAS,
+    'max': int,
+    'total': int
+    },required=True)
+
+################################################################
+
+class AliasSummary:
+
+    def __init__(self, json_file):
+        self.summary = get_alias_metrics(json_file)
+        self.max, self.total = get_max_total(self.summary)
+        self.validate()
+
+    def __repr__(self):
+        """A dict representation of alias summary."""
+
+        self.validate()
+        return self.__dict__
+
+    def __str__(self):
+        """A string representation of alias summary."""
+
+        return json.dumps({JSON_TAG: self.__repr__()}, indent=2)
+
+    def validate(self):
+        """Validate members of a summary object."""
+
+        return voluptuous.humanize.validate_with_humanized_errors(
+            self.__dict__, VALID_ALIAS_SUMMARY
+        )
+
+    def dump(self, filename=None, outdir=None):
+        """Write the alias summary to a file or stdout."""
+
+        util.dump(self, filename, outdir)
+
+    def render_report(self, filename=None, outdir=None):
+        """Write the alias summary to a file rendered as html."""
+
+        alias_html = templates.render_alias_summary(self)
+        util.dump(alias_html, filename or "alias.html", outdir)
+
+################################################################
+# Json key tags to access elements in cbmc json output
+
+JSON_POINTS_TO_SET_SIZE_KEY = 'PointsToSetSize'
+JSON_POINTER_KEY = 'Pointer'
+
+# Example cbmc json output
+# [
+#   {
+#     "program": "CBMC 5.13.0 (cbmc-5.13.1-44-g31340ca28)"
+#   },
+#   ...
+#   {
+#     "Pointer": "main::1::foo_p!0@1",
+#     "PointsToSet": [ ],
+#     "PointsToSetSize": 0,
+#     "RetainedValuesSet": [ ],
+#     "RetainedValuesSetSize": 0,
+#     "Value": "main::1::foo_p$object"
+#   },
+#   {
+#     "Pointer": "main::1::foo_p$object..next",
+#     "PointsToSet": [ "unknown" ],
+#     "PointsToSetSize": 1,
+#     "RetainedValuesSet": [ "unknown" ],
+#     "RetainedValuesSetSize": 1,
+#     "Value": "symex::invalid_object"
+#   },
+#   ...
+#   {
+#     "cProverStatus": "..."
+#   }
+# ]
+
+################################################################
+# Utility functions to generate alias summary
+
+def get_max_total(alias_summary):
+    max_val = 0
+    total = 0
+    for alias_item in alias_summary.items():
+        if int(alias_item[1][JSON_POINTS_TO_SET_SIZE_KEY]) > max_val:
+            max_val = int(alias_item[1][JSON_POINTS_TO_SET_SIZE_KEY])
+        total += int(alias_item[1][JSON_POINTS_TO_SET_SIZE_KEY]) * \
+            alias_item[1]['numOfDeref']
+
+    return max_val, total
+
+def get_alias_item(json_item):
+    pointer = json_item[JSON_POINTER_KEY]
+    json_item.pop(JSON_POINTER_KEY)
+    json_item.update({"numOfDeref": 1})
+
+    alias_item = {}
+    alias_item[pointer] = json_item
+
+    return alias_item
+
+def is_equal(summary, item):
+    for key in item.keys():
```

Comment:
What is this function doing?  It is not a predicate, because it has a side effect.  What is it checking for equality?

Reply:
The number of pointer dereferences is something that we calculate from the CBMC output. So dereferences are calculated as number of time CBMC refers to the points-to set for a given pointer. Every time CBMC refers to the points-to set, a corresponding `'pointer name, points-to set, points-to set size'` is displayed in CBMC json format. 
Here, we check if these fields match any existing parsed data, if yes, we increment number of dereferences, if not, we create a new data item in the dictionary.

Reply:
A pointer may have different points-to sets in different locations.  We are counting the number of times a pointer with a given points-to set is dereferenced.